spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/AggregatePath.java

```diff
@@ -0,0 +1,467 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.data.relational.core.mapping;
+
+import org.springframework.data.mapping.PersistentPropertyPath;
+import org.springframework.data.relational.core.sql.SqlIdentifier;
+import org.springframework.data.util.Lazy;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Represents a path within an aggregate starting from the aggregate root.
+ *
+ * @since 3.2
+ * @author Jens Schauder
+ */
+public class AggregatePath {
+
+	private final RelationalMappingContext context;
+
+	@Nullable private final RelationalPersistentEntity<?> rootType;
+
+	@Nullable private final PersistentPropertyPath<? extends RelationalPersistentProperty> path;
+	private final Lazy<SqlIdentifier> columnAlias = Lazy.of(() -> prefixWithTableAlias(getColumnName()));
+
+	public AggregatePath(RelationalMappingContext context,
+			PersistentPropertyPath<? extends RelationalPersistentProperty> path) {
+
+		Assert.notNull(context, "context must not be null");
+		Assert.notNull(path, "path must not be null");
+
+		this.context = context;
+		this.path = path;
+
+		this.rootType = null;
+	}
+
+	public AggregatePath(RelationalMappingContext context, RelationalPersistentEntity<?> rootType) {
+
+		Assert.notNull(context, "context must not be null");
+		Assert.notNull(rootType, "rootType must not be null");
+
+		this.context = context;
+		this.rootType = rootType;
+
+		this.path = null;
+	}
+
+	public static boolean isWritable(@Nullable PersistentPropertyPath<? extends RelationalPersistentProperty> path) {
+		return path == null || path.getLeafProperty().isWritable() && isWritable(path.getParentPath());
+	}
+
+	public boolean isRoot() {
+		return path == null;
+	}
+
+	/**
+	 * Tests if {@code this} and the argument represent the same path.
+	 *
+	 * @param path to which this path gets compared. May be {@literal null}.
+	 * @return Whence the argument matches the path represented by this instance.
+	 */
+	public boolean matches(@Nullable PersistentPropertyPath<RelationalPersistentProperty> path) {
```

Comment:
From the calling code, we should accept `AggregatePath` here as the nullable `PersistentPropertyPath` is an implementation detail.

Reply:
With `AggregatePath` as an argument this really is a `equals` so I turned it into `equals` and `hashCode`.